text_generation/causal_lm/cpp/continuous_batching/library/src/block_manager.hpp

```diff
@@ -141,16 +145,58 @@ class BlockManager {
         OPENVINO_ASSERT(m_block_table.erase(seq_id) == 1);
     }
 
-    bool can_append_slot(SequenceGroup::CPtr seq_group) {
-        // TODO: optimize this HEURISTIC
-        // it assumes that all sequences require new block, but maybe some of them
-        // don't share the same block
-        // let's count actual number of sequences, where last_block_id is the same
-        return seq_group->num_running_seqs() <= m_allocator.num_free_blocks();
+    void free_sequence_partially(size_t seq_id, size_t block_num) {
+        auto block_table = m_block_table[seq_id];
+
+        if (block_num == block_table.size())
+            return free_sequence(seq_id);
+
+        OPENVINO_ASSERT(block_table.size() >= block_num);
+        for (size_t idx = 0; idx < block_num; idx++) {
+            while (!block_table.back()->is_free()) {
+                m_allocator.free(block_table.back());
```

Comment:
sequence can free block only once

Reply:
Removed while here and added assert that block is free as agreed.